zaino-state/src/chain_index.rs

```diff
@@ -20,3 +20,65 @@ pub mod mempool;
 pub mod non_finalised_state;
 /// Common types used by the rest of this module
 pub mod types;
+
+use std::{collections::HashMap, sync::Arc};
+
+use futures::Stream;
+use non_finalised_state::{BlockchainSource, NonfinalizedBlockCacheSnapshot};
+pub use zebra_chain::parameters::Network as ZebraNetwork;
+use zebra_state::HashOrHeight;
+
+/// The combined index. Contains a view of the mempool, and the full
+/// chain state, both finalized and non-finalized, to allow queries over
+/// the entire chain at once. Backed by a source of blocks, either
+/// a zebra ReadStateService (direct read access to a running
+/// zebrad's database) or a jsonRPC connection to a validator.
+///
+/// TODO: Currently only contains the non-finalized state.
+pub struct ChainIndex {
+    // TODO: finalized state
+    // TODO: mempool
+    non_finalized_state: non_finalised_state::NonFinalizedState,
+}
+
+/// The interface to the chain index
+pub trait Index: Sized {
+    /// Creates a new chainindex from a connection to a validator
+    /// Currently this is a ReadStateService or JsonRpSeeConnector
+    async fn new<T: Into<BlockchainSource>>(
+        source: T,
+        network: ZebraNetwork,
+    ) -> Result<Self, non_finalised_state::InitError>;
+    /// Takes a snapshot of the non_finalized state. All NFS-interfacing query
+    /// methods take a snapshot. The query will check the index
+    /// it existed at the moment the snapshot was taken.
+    fn snapshot_nonfinalized_state(&self) -> Arc<NonfinalizedBlockCacheSnapshot>;
+    /// Given inclusive start and end indexes, stream all blocks
+    /// between the given indexes. Can be specified
+    /// by hash or height.
+    fn get_block_range<'snapshot: 'future, 'self_lt: 'future, 'future>(
+        &'self_lt self,
+        nonfinalized_snapshot: impl AsRef<NonfinalizedBlockCacheSnapshot> + Clone + 'snapshot,
+        start: Option<HashOrHeight>,
+        end: Option<HashOrHeight>,
+    ) -> Result<impl Stream<Item = Result<Vec<u8>, ()>> + 'future, ()>;
```

Comment:
The unit error type here doesn't tell us much about what can go wrong, maybe this is preferable?
```suggestion
    ) -> Result<impl Stream<Item = Result<Vec<u8>, Self::Error>> + 'future, ()>;
```
If it's a distinct set of errors that could be produced here, use a distinct associated type from `Self::Error`

Reply:
> The unit error type here doesn't tell us much about what can go wrong, maybe this is preferable?
> 
> If it's a distinct set of errors that could be produced here, use a distinct associated type from `Self::Error`

I think there will be a distinct set of errors that could be produced here, errors are all unit-typed as these error types will all be implementation defined...an associated error type for each method should be the way to go here.

Reply:
I would avoid having an associated type for each method; that ends up being a nightmare from a type ascription and constraint perspective. Forcing the trait implementer to create a single error type is not terrible.